cip-19: add eds identifier/container (#114)
celestiaorg/CIPs: cips/cip-19.md

This section defines all the supported Shwap messages, including share identifiers and containers. All the new
future messages should be described in this section.

#### EdsID

EdsID identifies the [DataSquare][square].

EdsID identifiers are formatted as shown below:

```text
EdsID {
    Height: u64;
}
```

The fields with validity rules that form EdsID are:

**Height**: A uint64 representing the chain height with the data square. It MUST be bigger than zero.

[Serialized](#serialization-for-share-identifiers) EdsID MUST have a length of 8 bytes.

#### Eds Container

> Eds containers encapsulate the [DataSquare][square]. Internally, they only keep the original data(1st quadrant) of the
> EDS with redundant data(2nd, 3rd and 4th quadrants) computable from the original data.

Reviewer: ```suggestion
Eds containers encapsulate the [DataSquare][square]. Internally, they only keep the original data (1st quadrant) of the
EDS with redundant data (2nd, 3rd and 4th quadrants) computable from the original data.
```

#### Row Container

> Row containers encapsulate the rows of the [DataSquare][square]. Internally, they only keep the left(original) half of
> the row with right(redundant) half recomputable from the left half.

Reviewer: ```suggestion
Row containers encapsulate the rows of the [DataSquare][square]. Internally, they only keep the left (original) half of
the row with right (redundant) half recomputable from the left half.
```

its peers. In contrast, the message might still be invalid according to the verification rules.

Bitswap still requires multihashes and CID codecs to be registered. Therefore, we provide a table for the
required [share identifiers](#share-identifiers) with their respective multihash and CID codec codes. This table
should be extended whenever any new share identifier or new version of an existing identifier is added.

| Name     | Multihash | Codec  |
|----------|-----------|--------|
| EdsID*   | N/A       | N/A    |
| RowID    | 0x7801    | 0x7800 |
| SampleID | 0x7811    | 0x7810 |
| DataID   | 0x7821    | 0x7820 |

*EdsID and container are excluded from Bitswap composition. Bitswap is limited to messages of size 256kb, while EDSes are

> expected to be bigger. Also, its more efficient to parallelize EDS requesting by rows.

Reviewer: ```suggestion
expected to be bigger. Also, it is more efficient to parallelize EDS requesting by rows.
```

its peers. In contrast, the message might still be invalid according to the verification rules.

Bitswap still requires multihashes and CID codecs to be registered. Therefore, we provide a table for the
required [share identifiers](#share-identifiers) with their respective multihash and CID codec codes. This table
should be extended whenever any new share identifier or new version of an existing identifier is added.

| Name     | Multihash | Codec  |
|----------|-----------|--------|
| EdsID*   | N/A       | N/A    |
| RowID    | 0x7801    | 0x7800 |
| SampleID | 0x7811    | 0x7810 |
| DataID   | 0x7821    | 0x7820 |

> *EdsID and container are excluded from Bitswap composition. Bitswap is limited to message of size 256kb, while EDSes are

Reviewer: ```suggestion
*EdsID and container are excluded from Bitswap composition. Bitswap is limited to messages of size 256kb, while EDSes are
```
